core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -0,0 +1,1001 @@
+                //
+                types = new SimpleImmutableEntry<>(null, (Class<?>) type);
+            }
+
+            return new AnnotatedValueResolver(annotation, httpElementName, pathVariable, shouldExist,
+                                              shouldWrapValueAsOptional, types.getKey(), types.getValue(),
+                                              defaultValue, resolver, aggregation);
+        }
+    }
+
+    private static boolean isFormData(@Nullable MediaType contentType) {
+        return contentType != null && contentType.belongsTo(MediaType.FORM_DATA);
+    }
+
+    enum AggregationStrategy {
+        NONE, ALWAYS, FOR_FORM_DATA;
+
+        /**
+         * Returns whether the request should be aggregated.
+         */
+        static boolean aggregationRequired(AggregationStrategy strategy, HttpRequest req) {
+            requireNonNull(strategy, "strategy");
+            switch (strategy) {
+                case ALWAYS:
+                    return true;
+                case FOR_FORM_DATA:
+                    return isFormData(req.headers().contentType());
+            }
+            return false;
+        }
+
+        /**
+         * Returns {@link AggregationStrategy} which specifies how to aggregate the request
+         * for injecting its parameters.
+         */
+        static AggregationStrategy resolve(List<AnnotatedValueResolver> resolvers) {
+            AggregationStrategy strategy = NONE;
+            for (final AnnotatedValueResolver r : resolvers) {
+                switch (r.aggregationStrategy()) {
+                    case ALWAYS:
+                        return ALWAYS;
+                    case FOR_FORM_DATA:
+                        strategy = FOR_FORM_DATA;
+                        break;
+                }
+            }
+            return strategy;
+        }
+    }
+
+    /**
+     * A context which is used while resolving parameter values.
+     */
+    static class ResolverContext {
+        private final ServiceRequestContext context;
+        private final HttpRequest request;
+
+        @Nullable
+        private final AggregatedHttpMessage message;
+
+        @Nullable
+        private volatile HttpParameters httpParameters;
+
+        ResolverContext(ServiceRequestContext context, HttpRequest request,
+                        @Nullable AggregatedHttpMessage message) {
+            this.context = requireNonNull(context, "context");
+            this.request = requireNonNull(request, "request");
+            this.message = message;
+        }
+
+        ServiceRequestContext context() {
+            return context;
+        }
+
+        HttpRequest request() {
+            return request;
+        }
+
+        @Nullable
+        AggregatedHttpMessage message() {
+            return message;
+        }
+
+        HttpParameters httpParameters() {
+            // Use a temporary variable to reduce the number of reads of the volatile field.
```

Comment:
Borrowed from https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/concurrent/LazyInitializer.java#L97

Reply:
I think it's too trivial and well known pattern (double-checked locking) and thus there's no need for an attribution.